### PR TITLE
 ci: run mir-opt tests on PR CI also as 32-bit (for `EMIT_MIR_FOR_EACH_BIT_WIDTH`).

### PR DESCRIPTION
**Pull request by eddyb**
Background: #69916 and [`src/test/mir-opt/README.md`](https://github.com/rust-lang/rust/blob/master/src/test/mir-opt/README.md):
> By default 32 bit and 64 bit targets use the same dump files, which can be problematic in the
presence of pointers in constants or other bit width dependent things. In that case you can add
>
> ```
> // EMIT_MIR_FOR_EACH_BIT_WIDTH
> ```
>
> to your test, causing separate files to be generated for 32bit and 64bit systems.

However, if you change the output of such a test (intentionally or not), or if you add a test and it varies between 32-bit and 64-bit platforms, you have to run this command (for a x64 linux host):
`./x.py test --stage 1 --target x86_64-unknown-linux-gnu --target i686-unknown-linux-gnu --bless  src/test/mir-opt`

Otherwise, bors trying to merge the PR will fail, since we test 32-bit targets there.
But we don't on PR CI, which means there's no way the PR author would know (unless they were burnt by this already and know what to look for).

This PR resolves that by running `mir-opt` tests for ~~`i686-unknown-linux-gnu`~~, on PR CI.
**EDIT**: switched to `armv5te-unknown-linux-gnueabi` to work around LLVM 7 crashes (see https://github.com/rust-lang/compiler-builtins/pull/311#issuecomment-612270089), found during testing.

cc @rust-lang/wg-mir-opt @rust-lang/infra

### CODE REVIEW

File: src/test/mir-opt/inline/inline-into-box-place.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // ignore-wasm32-bare compiled with panic=abort by default
 // compile-flags: -Z mir-opt-level=3
-// only-64bit FIXME: the mir representation of RawVec depends on ptr size
+// EMIT_MIR_FOR_EACH_BIT_WIDTH
```
Comment (Mark-Simulacrum) [inline src/test/mir-opt/inline/inline-into-box-place.rs:3]: Is this already implemented somewhere in a prior PR?

Can we also avoid shouting and use the same snake-case as is used for other compiletest options?

Comment (eddyb) [reply src/test/mir-opt/inline/inline-into-box-place.rs:3]: The annotation is from #69916 which was copying the style of the old `mir-opt` tests.

I've only added it here to get rid of the stray `only-64bit` (all other `mir-opt` tests that have platform differences use `// EMIT_MIR_FOR_EACH_BIT_WIDTH` already)

Comment (Mark-Simulacrum) [reply src/test/mir-opt/inline/inline-into-box-place.rs:3]: Ah, okay, then fine by me. I would like to see the shouting removed though, seems odd :)

**Comment by Mark-Simulacrum** [comment]
r=me when ready

File: src/test/mir-opt/inline/inline-into-box-place.rs (hunk 1 above)
Comment (eddyb) [reply src/test/mir-opt/inline/inline-into-box-place.rs:3]: Yeah, I agree, I think the `compiletest` precedent should be stronger than the old `mir-opt` testing (which #70721 already removed). Maybe someone from @rust-lang/wg-mir-opt wants to  play around with this.

(The implementation is also ad-hoc, it doesn't use the standard `compiletest` directives, we'd want to fix that too).

**Comment by eddyb** [comment]
Uhh why can we crash `rustc` like this?! Can I have the coredump?
> `rustc --crate-name compiler_builtins ...` (signal: 11, SIGSEGV: invalid memory reference)

**Comment by eddyb** [comment]
Could this be a LLVM7 bug? Since that's the only difference I see with the other Dockerfiles.

**Comment by Mark-Simulacrum** [comment]
Hm, I guess it could be, yeah. Let's randomly try @bors try and check that build too? None of the other builders that have run so far actually run LLVM I think

**Comment by eddyb** [comment]
For the record, here's what's actually happening (and why I couldn't find `llvm-config`):
* [`llvm-7-tools`](https://packages.ubuntu.com/bionic-updates/llvm-7-tools) depends on `llvm-7-dev` (and `llvm-8-*` is the same)
* [`llvm-9-dev`](https://packages.ubuntu.com/bionic-updates/llvm-9-dev) depends on `llvm-9-tools`

So we have to depend on `llvm-9-dev` if we ever try to use LLVM 9 on Ubuntu.

**Comment by eddyb** [comment]
We have confirmation: LLVM 9 fixes the `rustc` segfault while compiling `compiler_builtins`.
I'll try with LLVM 8 next.

<hr/>

But my test commit worked!
```diff
23	    }
24	}
25	
+	// HACK(eddyb) this is for testing if PR CI catches a 32-bit mir-opt failure.
+	
26	alloc0 (static: FOO, size: 8, align: 4) {
27	    ╾alloc24+0╼ 03 00 00 00                         │ ╾──╼....
28	}
```

I just have to include a commit from another branch that fixes the diff direction (should be `-` not `+`).

**Comment by eddyb** [comment]
Ah it's not `rustc` getting miscompiled, LLVM 7 just crashes whereas LLVM 8 prints this:
```
 cmpxchg instructions must be atomic.
  %10 = cmpxchg i64* %9, i64 0, i64 0 unordered not_atomic
in function __llvm_memcpy_element_unordered_atomic_8
LLVM ERROR: Broken function found, compilation aborted!
error: could not compile `compiler_builtins`.
```

**Comment by eddyb** [comment]
Oh but we expect to be able to run the tests somehow? Can I disable that :/?
(Or maybe find whatever allows other builds to run tests)

File: src/test/mir-opt/inline/inline-into-box-place.rs (hunk 1 above)
Comment (RalfJung) [reply src/test/mir-opt/inline/inline-into-box-place.rs:3]: Also is there documentation somewhere on the mir-opt format? I checked rustc-dev-guide, did not find anything.

**Comment by eddyb** [comment]
Added `--pass=build` support to `mir-opt` tests and used that to bypass having to execute anything.

**Comment by eddyb** [comment]
And, success on LLVM 7!
```
2020-04-11T12:21:59.9749761Z ---- [mir-opt] mir-opt/const_allocation2.rs stdout ----
2020-04-11T12:21:59.9749920Z 23     }
2020-04-11T12:21:59.9750017Z 24 }
2020-04-11T12:21:59.9750122Z 25 
2020-04-11T12:21:59.9750468Z - // HACK(eddyb) this is for testing if PR CI catches a 32-bit mir-opt failure.
2020-04-11T12:21:59.9750745Z - 
2020-04-11T12:21:59.9750920Z 28 alloc0 (static: FOO, size: 8, align: 4) {
2020-04-11T12:21:59.9751299Z 29     ╾alloc24+0╼ 03 00 00 00                         │ ╾──╼....
2020-04-11T12:21:59.9751552Z 
```

**Pull request merged**